File: ruby/raindrops/raindrops.rb
require 'prime'

class Raindrops
	VERSION = 1

	@@rain = {
		3 => 'Pling',
		5 => 'Plang',
		7 => 'Plong'
	}

	def self.convert(n)
		primes = Prime.prime_division(n).flatten.uniq.select { |n| @@rain.has_key?(n) }
		if primes.empty?
			n.to_s
		else
			primes.sort.map{ |n| @@rain[n] }.join("")
		end
	end
end

File: haskell/raindrops/src/Raindrops.hs
module Raindrops (convert) where

type Drop = (Int, String)

raindrops :: [Drop]
raindrops = [(3, "Pling"), (5, "Plang"), (7, "Plong")]

convert :: Int -> String
convert n = case filter hasDrop raindrops of
  [] -> show n
  xs -> concatMap snd xs
  where
    hasDrop (factor, _) = n `divisibleBy` factor
    divisibleBy x y = x `mod` y == 0
